modules/porous_flow/src/materials/PorousFlowFluidStateBaseMaterial.C

```diff
@@ -0,0 +1,197 @@
+              : &this->template declareProperty<std::vector<std::vector<Real>>>(
+                    "dPorousFlow_fluid_phase_internal_energy" + _ext + "_dvar"))
+{
+  // Set the size of the FluidStateProperties vector
+  _fsp.resize(_num_phases, FluidStateProperties(_num_components));
+}
+
+template <>
+GenericReal<false>
+PorousFlowFluidStateBaseMaterialTempl<false>::genericValue(const ADReal & value)
+{
+  return MetaPhysicL::raw_value(value);
+}
+
+template <>
+GenericReal<true>
+PorousFlowFluidStateBaseMaterialTempl<true>::genericValue(const ADReal & value)
+{
+  return value;
+}
+
+template <bool is_ad>
+void
+PorousFlowFluidStateBaseMaterialTempl<is_ad>::initQpStatefulProperties()
+{
+  _is_initqp = true;
+  // Set the size of pressure and saturation vectors
+  PorousFlowVariableBaseTempl<is_ad>::initQpStatefulProperties();
+
+  // Set the size of all other vectors
+  setMaterialVectorSize();
+
+  thermophysicalProperties();
+
+  // Set the initial values of the properties at the nodes.
+  // Note: not required for qp materials as no old values at the qps are requested
+  // unless the material is AD (for the FV case there are no nodal materials)
+  if (_nodal_material || is_ad)
+    for (unsigned int ph = 0; ph < _num_phases; ++ph)
+    {
+      _saturation[_qp][ph] = genericValue(_fsp[ph].saturation);
+      _porepressure[_qp][ph] = genericValue(_fsp[ph].pressure);
+      _fluid_density[_qp][ph] = genericValue(_fsp[ph].density);
+      _fluid_viscosity[_qp][ph] = genericValue(_fsp[ph].viscosity);
+      _fluid_enthalpy[_qp][ph] = genericValue(_fsp[ph].enthalpy);
+      _fluid_internal_energy[_qp][ph] = genericValue(_fsp[ph].internal_energy);
+
+      for (unsigned int comp = 0; comp < _num_components; ++comp)
+        _mass_frac[_qp][ph][comp] = genericValue(_fsp[ph].mass_fraction[comp]);
+    }
+}
+
+template <bool is_ad>
+void
+PorousFlowFluidStateBaseMaterialTempl<is_ad>::computeQpProperties()
+{
+  _is_initqp = false;
+  // Prepare the derivative vectors
+  PorousFlowVariableBaseTempl<is_ad>::computeQpProperties();
+
+  // Set the size of all other vectors
+  setMaterialVectorSize();
+
+  // Calculate all required thermophysical properties
+  thermophysicalProperties();
+
+  for (unsigned int ph = 0; ph < _num_phases; ++ph)
+  {
+    _saturation[_qp][ph] = genericValue(_fsp[ph].saturation);
+    _porepressure[_qp][ph] = genericValue(_fsp[ph].pressure);
+    _fluid_density[_qp][ph] = genericValue(_fsp[ph].density);
+    _fluid_viscosity[_qp][ph] = genericValue(_fsp[ph].viscosity);
+    _fluid_enthalpy[_qp][ph] = genericValue(_fsp[ph].enthalpy);
+    _fluid_internal_energy[_qp][ph] = genericValue(_fsp[ph].internal_energy);
+
+    for (unsigned int comp = 0; comp < _num_components; ++comp)
+      _mass_frac[_qp][ph][comp] = genericValue(_fsp[ph].mass_fraction[comp]);
+  }
+}
+
+template <bool is_ad>
+void
+PorousFlowFluidStateBaseMaterialTempl<is_ad>::setMaterialVectorSize() const
+{
+  _fluid_density[_qp].assign(_num_phases, 0.0);
```

Comment:
since you're going to fill those, you dont need this assign, you could have resize everywhere instead of assign and save the "setting to 0" operation

Reply:
kind of a micro-optimization though. Porbably not needed

Reply:
I'll probably change it anyway